src/crimson/admin/admin_socket.cc

```diff
@@ -0,0 +1,896 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2011 New Dream Network
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#include "common/version.h"
+#include "crimson/net/Socket.h"
+// activate for some inline unit-testing: #define UNIT_TEST_OSD_ASOK
+#include "crimson/admin/admin_socket.h"
+#include "crimson/common/log.h"
+//#include "seastar/testing/test_case.hh"
+#include "seastar/net/api.hh"
+#include "seastar/net/inet_address.hh"
+#include "seastar/core/reactor.hh"
+#include "seastar/core/thread.hh"
+//#include "seastar/util/log.hh"
+#include "seastar/util/std-compat.hh"
+#include <boost/algorithm/string.hpp>
+
+/*!
+  A Crimson-wise version of the admin socket - implementation file
+
+  \todo handle the unlinking of the admin socket. Note that 'cleanup_files' at-exit functionality is not yet
+        implemented in Crimson
+*/
+
+namespace {
+  seastar::logger& logger() {
+    return ceph::get_logger(ceph_subsys_osd);
+  }
+}
+
+#ifdef UNIT_TEST_OSD_ASOK
+namespace asok_unit_testing {
+  seastar::future<> utest_run_1(AdminSocket* asok);
+}
+#endif
+
+/*!
+  Hooks table - iterator support
+
+  Note: Each server-block is locked (via its gate member) before
+        iterating over its entries.
+ */
+
+AdminHooksIter AdminSocket::begin() {
+  AdminHooksIter it{*this};
+  it.m_miter->second.m_gate.enter();
```

Comment:
I guess the idea here is that a user of AdminHooksIter might defer to a future while holding a reference to the iterator and you want to avoid the hooks mutating during that time.  I'd suggest that we don't really need an atomic view of the underlying container, would it be possible to modify this so that "concurrent" modifications are tolerated without a lock?

Another option would be to iterate over all of the hooks synchronously and then wait on the result of all of the futures at the end -- that would eliminate the possibility of concurrent modification.

Keep in mind, even once we go to multiple reactors, each reactor will have its own set of hooks, so we're not worried about true concurrent mutation.

Reply:
No, I guess the intention is to ensure that two concurrent admin socket commands don't hit the same hook at the same time?

Reply:
To answer the questions regarding the locking decisions:
execute_line() uses parse_cmd() to analyze the incoming message and locate the command in one of the servers' blocks. Once the relevant registered API is located, the general RW-lock on the servers table is released. The only thing that keeps the server, the one that actually handles that registered API, from disappearing from under our feet is the gate we've entered.
Now - the code can be made much simpler if we simply keep the general RW lock at all times while executing an asok command. I can then remove all gates-related handling. I am not sure that that's a bad design decision. It's just that when I wrote the code - the gates came first, as I was trying to avoid the non-granular RW lock.